### PR TITLE
Portilla simoncelli notebook

### PR DESCRIPTION
**Pull request by kbonnen**
This is the replication of Portilla & Simoncelli 2000

### CODE REVIEW

**Comment by kbonnen** [comment]
- [ ] Need to add image examples to OSF.
- [ ] Need to add download code for the image examples used here.

**Comment by lyndond** [comment]
I left some comments in the NBReviewer too.

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 1]
```diff
@@ -333,16 +333,19 @@ def forward(self, image, scales=None):
...
+                if s not in scales:
+                    representation_vector[:,:,ii]=0
+            # ind = torch.tensor(
+            #     [
+            #         i
+            #         for i, s in enumerate(self.representation_scales)
+            #         if s in scales
+            #     ]
+            # ).to(image.device)
+            # return representation_vector.index_select(-1, ind)
```
Comment (billbrod) [inline plenoptic/simulate/models/portilla_simoncelli.py:348]: What's going on here? It looks like you changed the behavior with `scales` so that you zero values out instead of just returning the relevant values. Why'd you make that change? (and you should delete the old way, not just comment it out!)

**Review by billbrod — CHANGES_REQUESTED** [review]
I've got a question about the change you made to `scales`.

Also:
- need to upload the required images, like you said
- I just realized `im_shape`, during initialization isn't documented. Can you do that? In particular, it should specify that `im_shape` should be 2d, giving the height and width (rather than 4d)

I'll look at the notebook in a sec.

**Comment by billbrod** [comment]
Many comments on the notebook. The text trailed off in sections 3 and 4, so that needs to be addressed.

Also a more general point about notebooks: you have several large blocks of code that get used in the notebook. I'm not a huge fan of large blocks of un-described code. we could move them into a helper script somewhere so that they're just loaded in here. advantage is that we don't have large blocks of code that might cause users to zone out and then it's easier to keep track of changes to them, but then we're giving ourselves a piece of code that's only used in one place and isn't really meant for users (right?).

what do people think, how should we handle this?

**Comment by kbonnen** [comment]
<p>I have moved this up as suggested.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993340'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Updated text and grouped the images by their "type" as you've described.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993356'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>references updated.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993371'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>You and I had a long conversation about the large blocks of code before I decided to put them here.  As you point out, this code is only meant for this demonstration.  We don't really mean it for users.  On the other hand it's a bit distracting from the point in these sections.  I'm happy to pull it out if that's preferable.</p>

**Comment by kbonnen** [comment]
<p>I moved up the visual textures section as suggested.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993280'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Yep.  As a result of your suggestions this is no longer in the first section.  We used to have a whole introduction to metamers in this notebook and when it got cut the beginning of this notebook got weird.  Your other comments have helped to resolve that.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993296'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Substituted for the wicker basket texture.  The model does a great job at this.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993323'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>the flowers were just always my favorite...</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993323'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>This seems out of place now.  I've actually cut it.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993392'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Other than showing an example... is there something you want me to do to show it does a good job?  It doesn't run fast enough to easily do multiple examples. </p><p> </p><p>We could pre-calculate a bunch of examples. I'd sort of prefer not to, but that's mostly because it's additional work.  If you really want this, I will do it.</p><p> </p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993417'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>See conversation in later comments.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993434'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>good catch.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993482'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Removed.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993611'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>I don't think this belongs in an entirely separate notebook.  If it doesn't go here, then it goes in the docstring of the Portilla Simoncelli model.</p><p> </p><p>I've made it less defeatist sounding :) </p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993659'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Turns out I'm missing all the images from figure 12.  I bet this was a place holder...</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993584'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Added some explanatory text and an excerpt from the paper and an image showing the masked out area.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993643'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>I have added some explanatory text.  Will you take a look?  Do you want another level of a description?  Like a description for each image. (Note: in many cases this will be a bit of a guess since this is not included in Eero's dataset).</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993566'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>I've added some help text to the plot-synthesis-status output.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993323'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>done.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993454'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Ok.  So I have screen grabbed these images.  They are blurrier than the what the originals were... but I also think the python implementation may not be as good at these black and white images anyways (we end up with more blurry shadows).</p><p> </p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993584'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: plenoptic/simulate/models/portilla_simoncelli.py (hunk 1 above)
Comment (kbonnen) [reply plenoptic/simulate/models/portilla_simoncelli.py:348]: There's an issue with converting to and from a dictionary if you actually remove the values.  I don't remember all the details.  I have removed the comments.

**Comment by kbonnen** [comment]
> * [x]  Need to add image examples to OSF.
> 
>     * [ ]  Need to add download code for the image examples used here.

@billbrod How do you recommend doing the download code in the notebook?  I can't find any examples in quick search of MAD and steerable pyramid and a few others.

**Comment by kbonnen** [comment]
@billbrod Can we find 15-20 minutes this week? https://calendly.com/kbonnen/half-hour-meeting-clone  I have a few open questions that I need to address but I need some feedback from you.  We can also decide on how we want to proceed with getting this notebook.  I'd like to have it done by the end of this week.

**Comment by kbonnen** [comment]
<p>Billy says I don't have to make more examples.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993417'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Fixed.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993499'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>We're going to put these functions into hidden cells (and note that they are there) after we document them.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993628'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by kbonnen** [comment]
<p>Add notation to docstring that references notebook for issue 1 and 2 mentioned here.</p>

---

 <a href='https://app.reviewnb.com/LabForComputationalVision/plenoptic/pull/175/discussion/#comment-1246993659'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by billbrod** [comment]
A question here: we have a couple large convenience functions that are created in this notebook. I don't love having it, because it looks dense but isn't worth explaining in detail. Does anyone have an idea for what to do with it? We tried hiding the cells, and that doesn't work. We could also create them in a neighboring script, but that doesn't seem much cleaner...

**Review by billbrod — APPROVED** [review]
Looks good to Kate and I

**Review by Bichidian — CHANGES_REQUESTED** [review]

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 2]
```diff
@@ -333,16 +333,11 @@ def forward(self, image, scales=None):
...
-                [
-                    i
-                    for i, s in enumerate(self.representation_scales)
-                    if s in scales
-                ]
-            ).to(image.device)
-            return representation_vector.index_select(-1, ind)
+            for ii,s in enumerate(self.representation_scales):
+                if s not in scales:
+                    representation_vector[:,:,ii]=0
```
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/portilla_simoncelli.py:340]: I'm actually going to revert this -- I think removing the stats is a more obvious way to do it (and I rely on it for validation in #196)

**Review by Bichidian — APPROVED** [review]
